### PR TITLE
Filter out groups that are still running from real failures

### LINKED ISSUE #697: If a group is still running on a parent push, don't consider a regression as new
E.g. in case of backfills, there is a chance a group will finish on a child push before a parent push.
If there is a failure happening on a push, and the same group is still running on the parent, then we can't tell if the failure is new or pre-existing.

Not sure we can identify this situation though, as the task running the group is still running, and so errorsummary.log is not available yet.

### PR DESCRIPTION
**Pull request by EvaBardou**
Closes #697

### CODE REVIEW

**Comment by EvaBardou** [comment]
@marco-c I'm not sure how we can test this though, if you have any idea, don't hesitate :smile:

File: mozci/push.py [hunk 1]
```diff
@@ -533,6 +535,39 @@ def scheduled_duration(self):
...
+        """Checks if the provided group is still running on this push.
+
+        Returns:
+            bool: True, if the group is still running.
+                  False, if the group is completed.
+        """
+        running_tasks = [
+            task
+            for task in self.tasks
+            if task.state not in ["completed", "failed", "exception"]
```
Comment (marco-c) [inline mozci/push.py:548]: This makes me think that we should update `completed_cached_tasks` to also consider `failed` and `exception`, we were only caching `completed` :)

File: mozci/push.py [hunk 2]
```diff
@@ -533,6 +535,39 @@ def scheduled_duration(self):
...
+                ):
+                    return True
+        else:
+            group_types = set([task.tags.get("tests-type", "") for task in group.tasks])
+            running_types = set(
+                [task.tags.get("tests-type", "") for task in running_tasks]
+            )
+            return any(
+                tests_type in running_types for tests_type in group_types if tests_type
+            )
```
Comment (marco-c) [inline mozci/push.py:567]: Nit: this could be simplified by doing something like:
```Python
return not group_types.isdisjoint(running_types)
```
or
```Python
return len(group_types | running_types) != 0
```

File: mozci/push.py [hunk 3]
```diff
@@ -533,6 +535,39 @@ def scheduled_duration(self):
...
+
+        return False
```
Comment (marco-c) [inline mozci/push.py:569]: Nit: move this in the first `if` branch after the for, as the else branch always returns

File: mozci/push.py [hunk 4]
```diff
@@ -1181,10 +1219,15 @@ def classify_regressions(
...
+        )
 
-        # Real failure are groups with likely regressions that were selected with high confidence
-        # AND failing across config
-        real_failures = groups_regressions & groups_relevant_failure & groups_high
+        # Real failures are groups that finished running, with likely regressions that were selected
+        # with high confidence AND failing across config
+        real_failures = (
+            groups_regressions & groups_relevant_failure & groups_high
+        ) - groups_still_running
```
Comment (marco-c) [inline mozci/push.py:1230]: We should move these still running groups to the unknown failures list (but take into account that we are not going to do any actions on them as part of #654.

File: mozci/task.py [hunk 5]
```diff
@@ -220,6 +220,10 @@ def is_backfill(self) -> bool:
     def is_retrigger(self) -> bool:
         return self.tags.get("action", "").startswith("retrigger-")
 
+    @property
+    def is_tests_grouped(self) -> bool:
```
Comment (marco-c) [inline mozci/task.py:224]: Nit: could you use this function in the `label_summaries` function in push.py too?

File: mozci/task.py [hunk 6]
```diff
@@ -303,7 +307,8 @@ def backfill(self, push):
                 "taskGroupId": decision_task.id,
                 "input": {
                     "times": 5
-                    if self.classification == "not classified" or self.classification in INTERMITTENT_CLASSES
+                    if self.classification == "not classified"
+                    or self.classification in INTERMITTENT_CLASSES
```
Comment (marco-c) [inline mozci/task.py:311]: Nit: could you revert this unnecessary change?

**Comment by marco-c** [comment]
> @marco-c I'm not sure how we can test this though, if you have any idea, don't hesitate smile

You could ask @Archaeopteryx via chat to tell you when sheriffs backfill a task for reftest and one for mochitest, then we can try running the `is_group_running` function for one of the groups of that task and check that it is returning True until the backfill finishes (and False since then).

File: mozci/push.py (hunk 1 above)
Comment (EvaBardou) [reply mozci/push.py:548]: Would you like me to make the update in this PR or elsewhere?

Comment (marco-c) [reply mozci/push.py:548]: You could do it in this PR, defining this list in the defs.py file

File: mozci/task.py (hunk 6 above)
Comment (marco-c) [reply mozci/task.py:311]: Weird that it doesn't fail on master :S
Anyway, I've run it locally and pushed the reformatted code to master, so you can rebase.

File: mozci/push.py [hunk 7]
```diff
@@ -533,6 +535,44 @@ def scheduled_duration(self):
...
+                    return True
+            return False
+
+        group_types = set(
+            [
+                task.tags["tests-type"]
+                for task in group.tasks
+                if task.tags.get("tests-type")
+            ]
+        )
```
Comment (marco-c) [inline mozci/push.py:566]: Nit: no need to create a list first:
```Python
        group_types = set(
                task.tags["tests-type"]
                for task in group.tasks
                if task.tags.get("tests-type")
        )
```
or
```Python
        group_types = {
                task.tags["tests-type"]
                for task in group.tasks
                if task.tags.get("tests-type")
        }
```

File: mozci/push.py [hunk 8]
```diff
@@ -533,6 +535,44 @@ def scheduled_duration(self):
...
+        running_types = set(
+            [
+                task.tags["tests-type"]
+                for task in running_tasks
+                if task.tags.get("tests-type")
+            ]
+        )
```
Comment (marco-c) [inline mozci/push.py:573]: Nit: same here, no need to create a list first

**Comment by EvaBardou** [comment]
@marco-c A couple of tests are failing in CI due to this change https://github.com/mozilla/mozci/pull/791/files#diff-cbdcefec12b605b952245761dbfbbbb2ec7fc73addcc510c2b8dc48a60bd5d02R345, I can reproduce locally, it is altering the behavior of the `classify` function, what should we do about this?

File: mozci/push.py [hunk 9]
```diff
@@ -1169,6 +1205,9 @@ def classify_regressions(
             for g in all_groups
             if g.name not in list(bugbug_selection["groups"].keys())
         }
+
+        groups_still_running = {g.name for g in all_groups if self.is_group_running(g)}
```
Comment (marco-c) [inline mozci/push.py:1209]: We should actually check if the group is still running in any of the parents of the push.
We iterate the parents and we either stop if we find the group run, or add the group to the groups_still_running set.
Pseudocode:
```Python
for group in real_failures:
    for parent in push.parents:
        # If the group run in a parent, then we don't care if it's still running in a grandparent.
        if group in parent.group_summaries:
            break

        if parent.is_group_running(group):
            groups_still_running.add(group)
            break

real_failures -= groups_still_running
unknown_failures |= groups_still_running
```

**Comment by EvaBardou** [comment]
> Are the tests still failing after my latest changes?

Unfortunately, yes

> If yes, we should make sure in the mocks we are not considering any group as still running, then the tests should behave exactly the same.

We aren't, I patched the function to always return `False` with: `monkeypatch.setattr(Push, "is_group_running", lambda *args: False)`
But if a group isn't running, currently it is added to `unknown_failures`, thus changing the classified failures and the overall classification of the Push in some cases.

**Comment by marco-c** [comment]
Let's see if it still happens after https://github.com/mozilla/mozci/pull/791#discussion_r902901661, but it shouldn't happen if you mock all groups to consider them as finished
